subprojects/core/src/main/java/org/gradle/api/internal/file/copy/CopyActionExecuter.java

```diff
@@ -40,6 +40,7 @@ public CopyActionExecuter(Instantiator instantiator, ObjectFactory objectFactory
     public WorkResult execute(final CopySpecInternal spec, CopyAction action) {
         final CopyAction effectiveVisitor = new DuplicateHandlingCopyActionDecorator(
                 new NormalizingCopyActionDecorator(action, fileSystem),
+                spec,
```

Comment:
`DuplicateHandlingCopyActionDecorator` now stores the `CopySpec` to use the destination dir in the failure message upon hitting duplicate files.